ckan/views/__init__.py

```diff
@@ -62,6 +63,10 @@ def set_cache_control_headers_for_response(response: Response) -> Response:
     else:
         response.cache_control.private = True
 
+    # Invalidate cached pages upon login/logout
+    if limit_cache_by_cookie:
+        response.vary.add("Cookie")
+
```

Comment:
This should be in the if: block above because otherwise caching is disabled, right?

Reply:
No. As of CKAN 2.10, "disabled" isn't really disabled; it's just "private to the browser". Which is secure, but doesn't fix the problem of pages being cached beyond logout.

Reply:
Is CSRF going to generally be applicable for non-logged in users? Or rather, do we set that cookie for pages where there's no form for non-logged in users?  If we do set a csrf cookie on everything, then this essentially blows caching for all html responses. 

In general, I think there are three cases:

* Caching Allowed (public pages, assets, especially the home page when not logged in)
* Private caching allowed (e.g. logged in + back button, on most of the site)
* Caching disabled (login/logout redirect targets)

Reply:
> If we do set a csrf cookie on everything, then this essentially blows caching for all html responses.

If there is a constantly changing CSRF token in the forms, then caching them will break the site. You'll have users submitting stale tokens and getting errors.

It's okay if the token is generated once per session, or once every 30 minutes. But if it's created anew for each request, then caching is not an option.

> Caching Allowed (public pages, assets, especially the home page when not logged in).

As soon as you *do* log in, though, every cached page needs to be invalidated for you.

> Private caching allowed (e.g. logged in + back button, on most of the site)

This needs to be invalidated as soon as you log out. 

> Caching disabled (login/logout redirect targets)

Well, those should have been POST requests all along, which takes care of it.